rfcs/20200612-stream-executor-c-api.md

```diff
@@ -0,0 +1,557 @@
+            on developer experience and feedback. When the APIs become more mature, 
+            we will try to keep them stable (in a best-effort manner) until the new 
+            TensorFlow stack is available.
+
+# Design Proposal
+
+## StreamExecutorInterface
+
+[StreamExecutorInterface](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/stream_executor/stream_executor_internal.h;l=166?q=StreamExecutorinterface)
+is quite large and some of its methods are only sporadically used. Therefore, we
+plan to wrap only a subset of key StreamExecutorInterface functionality.
+
+See proposed C API below:
+
+```cpp
+#include <stddef.h>
+#include <stdint.h>
+
+#define SE_MAJOR 0
+#define SE_MINOR 0
+#define SE_REVISION 1
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef SE_Stream_st* SE_Stream;
+typedef SE_Event_st* SE_Event;
+typedef SE_Timer_st* SE_Timer;
+typedef TF_Status* (*TF_StatusCallbackFn)(void*);
+
+#ifndef TF_BOOL_DEFINED
+#define TF_BOOL unsigned char
+#endif // TF_BOOL_DEFINED
+
+#ifndef TF_OFFSET_OF_END
+#define TF_OFFSET_OF_END(TYPE, MEMBER) (offsetof(TYPE, MEMBER) + sizeof(((TYPE *)0)->MEMBER))
+#endif // TF_OFFSET_OF_END
+
+typedef struct SE_PlatformId {
+  size_t struct_size;
+  void* ext;
+  void* id;  // aka stream_executor::Platform::Id
+} SE_PlatformId;
+
+#define SE_PLATFORMID_STRUCT_SIZE TF_OFFSET_OF_END(SE_PlatformId, id)
+
+typedef struct SE_TimerFns {
+  size_t struct_size;
+  void* ext;
+  uint64_t (*nanoseconds)(SE_Timer timer);
+  uint64_t (*microseconds)(SE_Timer timer);
+} SE_TimerFns;
+
+#define SE_TIMER_FNS_STRUCT_SIZE TF_OFFSET_OF_END(SE_TimerFns, microseconds)
+
+typedef struct SE_AllocatorStats {
+  size_t struct_size;
+  void* ext;
+  int64_t num_allocs;
+  int64_t bytes_in_use;
+  int64_t peak_bytes_in_use;
+  int64_t largest_alloc_size;
+
+  int8_t has_bytes_limit;
+  int64_t bytes_limit;
+
+  int64_t bytes_reserved;
+  int64_t peak_bytes_reserved;
+
+  int8_t has_bytes_reservable_limit;
+  int64_t bytes_reservable_limit;
+
+  int64_t largest_free_block_bytes;
+} SE_AllocatorStats;
+
+#define SE_ALLOCATORSTATS_STRUCT_SIZE TF_OFFSET_OF_END(SE_AllocatorStats, largest_free_block_bytes)
+
+typedef enum SE_EventStatus {
+  SE_EVENT_UNKNOWN,
+  SE_EVENT_ERROR,
+  SE_EVENT_PENDING,
+  SE_EVENT_COMPLETE,
+} SE_EventStatus;
+
+typedef struct SE_Options {
+  size_t struct_size;
+  void* ext;
+  int32_t ordinal;
+} SE_Options;
+
+#define SE_OPTIONS_STRUCT_SIZE TF_OFFSET_OF_END(SE_Options, ordinal)
+
+typedef struct SE_Device {
+  size_t struct_size;
+  void* ext;
+  const char* name;
+  size_t name_len;
+
+  // Device vendor can store handle to their device representation
+  // here.
+  void* device_handle;
+
+  // Any kind of data that plugin device might want to store.
+  void* data;
+} SE_Device;
+
+#define SE_DEVICE_STRUCT_SIZE TF_OFFSET_OF_END(SE_Device, data)
+
+typedef struct SE_StreamExecutor {
+  size_t struct_size;
+  void* ext;
```

Comment:
Does the plugin need to provide all these function pointers?  Or can it provide a subset?

Reply:
This was determined the minimal required subset according to @rjpower

Reply:
Ok, maybe this should be explicitly stated?  (Apologies if this is already stated and I just missed it.)

Reply:
Added a sentence to clarify this:
We decided on this subset based on the PluggableDevice usecase as well as potential future devices such as TPUs.